### PR TITLE
add cardano address validation

### PR DESCRIPTION
**Pull request by refi93**

### CODE REVIEW

File: src/apps/cardano/sign_tx.py [hunk 1]
```diff
@@ -184,6 +188,8 @@ def _process_outputs(self):
                     raise wire.ProcessError(
                         "Each output must have address or address_n field!"
                     )
+                if validate_address(output.address) is False:
```
Comment (ppershing) [inline src/apps/cardano/sign_tx.py]: why not simply `if not validate_address():` ?

File: src/apps/cardano/address.py [hunk 2]
```diff
@@ -16,12 +25,24 @@ def derive_address_and_node(keychain, path: list):
...
+        address_unpacked = cbor.decode(address_hex)
+    except ValueError as e:
+        if __debug__:
+            log.exception(__name__, e)
+        return False
+
+    if not isinstance(address_unpacked, list) or len(address_unpacked) != 2:
+        return False
+
+    address_data_encoded = address_unpacked[0]
```
Comment (ppershing) [inline src/apps/cardano/address.py]: We should also check that `address_data_encoded` is a byte array

File: src/apps/cardano/sign_tx.py (hunk 1 above)
Comment (refi93) [reply src/apps/cardano/sign_tx.py]: fixed

File: src/apps/cardano/address.py (hunk 2 above)
Comment (refi93) [reply src/apps/cardano/address.py]: fixed

File: src/apps/cardano/address.py [hunk 3]
```diff
@@ -16,12 +25,27 @@ def derive_address_and_node(keychain, path: list):
...
-    )
-    return (address, node)
+    return (_encode_address_raw(address_data_encoded), node)
+
+
+def validate_address(address) -> bool:
+    try:
+        address_hex = base58.decode(address)
+        address_unpacked = cbor.decode(address_hex)
+    except ValueError as e:
```
Comment (ppershing) [inline src/apps/cardano/address.py]: Just a check -- both base58 and cbor throw only `ValueError`?

File: src/apps/cardano/address.py [hunk 4]
```diff
@@ -16,12 +25,27 @@ def derive_address_and_node(keychain, path: list):
...
-    address = base58.encode(
-        cbor.encode(
-            [cbor.Tagged(24, address_data_encoded), crc.crc32(address_data_encoded)]
-        )
-    )
-    return (address, node)
+    return (_encode_address_raw(address_data_encoded), node)
+
+
+def validate_address(address) -> bool:
```
Comment (ppershing) [inline src/apps/cardano/address.py]: I am not sure I like `validate_` prefix. Probably `is_valid_` is better but I can't find a good wording for it.
Also, this *does* not say what is being validated. What we validate is *only* outer CBOR encoding, not full validity.
Maybe `address_has_valid_root_cbor_structure` ?

**Pull request merged**